Joy/mpc server

joyqvq opened the pull request:
## Description 

Describe the changes or additions included in this PR.

## Test plan 

How did you test the new or updated feature?

joyqvq commented on crates/crypto/src/lib.rs:
```diff
@@ -298,6 +298,133 @@ pub fn create_full_id(package_id: &[u8; 32], id: &[u8]) -> Vec<u8> {
 }
 
+/// MPC-specific seal encrypt function.
+pub fn seal_encrypt_mpc(
```
review this

joyqvq commented on crates/crypto/src/lib.rs:
```diff
@@ -298,6 +298,133 @@ pub fn create_full_id(package_id: &[u8; 32], id: &[u8]) -> Vec<u8> {
+}
+
+/// MPC-specific seal decrypt function.
+pub fn seal_decrypt_mpc(
```
review this

joyqvq commented on crates/dkg-cli/README.md:
```diff
@@ -0,0 +1,398 @@
+# DKG CLI Tool
```
review the overall steps

joyqvq commented on crates/key-server/src/tests/e2e.rs:
```diff
@@ -386,6 +392,160 @@ async fn test_e2e_imported_key() {
 
+#[traced_test]
+#[tokio::test]
+async fn test_e2e_mpc() {
```
this test uses the real result from dkg-cli run (see readme for test data artifacts). and run all the move txn (for init, register, propose, approve, finalize). skim to review overall steps

benr-ml commented on move/committee/sources/committee.move:
```diff
@@ -0,0 +1,640 @@
+) {
+    assert!(committee.members.contains(&ctx.sender()), ENotMember);
+    let mut key_server = transfer::public_receive(&mut committee.id, ks);
+    key_server::update_url(&mut key_server, url, ctx);
```
it should be of the partial ks, no?

benr-ml commented on move/committee/sources/committee.move:
```diff
@@ -0,0 +1,640 @@
+    key_server::update_url(&mut key_server, url, ctx);
+    transfer::public_transfer(key_server, committee.id.to_address());
+}
+
```
let's add a TODO to handle package upgrades given an approval of threshold of the members

benr-ml commented on move/seal/sources/key_server.move:
```diff
@@ -24,62 +25,222 @@ public struct KeyServer has key, store {
+    key_type: u8, // For both types
+    pk: vector<u8>,  // For both types
+    server_type: ServerType,
+    threshold: u16,  // t for Committee, 0 for Single
```
let's move threshold to Committee enum

benr-ml replied:
maybe the url to Independent?

joyqvq replied:
good point, revised

benr-ml commented on crates/dkg-cli/README.md:
````diff
@@ -0,0 +1,398 @@
+Each party generates their ECIES and signing keypairs.
+
+```bash
+cargo run --bin dkg-cli generate-keys
````
or, should we wrap those steps in a python script which will handle all those parameters? (we can do that later, just wondering how much we can simplify things)

joyqvq replied:
you mean, a python code that do the key gen and move call for register at the same time? yeah we can simplify in the future

joyqvq commented on move/committee/sources/committee.move:
```diff
@@ -0,0 +1,640 @@
+            // Transition from Init to PreDKG
+            committee.state = State::PreDKG;
+            // Store candidate data as dynamic field
+            df::add(&mut committee.id, sender, CandidateData { enc_pk, signing_pk });
```
do we need the candidate data anymore after this state right? i originally keep it in df instead of table/map is that a table needs to be explicitly dropped.

benr-ml reviewed (commented):
the Move code looks good, left small comments. let's merge the unrelated changes to fastcrypto (updated deps, etc) and move to focus on the next component

joyqvq commented:
retargeted to isolate all unrelated changes to a different pr. i squashed the commits to make rebase a bit easier. changes are
- removed approve_committee and use propose and propose_for_rotation instead
- seal_test pkg so the current seal pkg is untouched
- added a few more move tests and condition checks. 

a few qqs: 
- init_committee_for_rotation: should we check the continuing members meets old committee threshold?
- do we want to do any validation for enc_pk, signing_pk, partial_pks and pk to be g2 elements? (since if they are not valid, the member should not approve, so easier to just keep the primitive type)
- should we enforce a max members to the committee? 20?

joyqvq commented on move/seal_testnet/sources/key_server.move:
```diff
@@ -0,0 +1,352 @@
+
+/// KeyServerV2: supports both single and committee-based key servers
+public struct KeyServerV2 has key, store {
+    id: UID,
```
to add df to keyserverV2, and an UID is needed 

```
        df::add(&mut v2.id, members[i], partial_key_server);
````

joyqvq commented on move/seal_testnet/sources/key_server.move:
```diff
@@ -0,0 +1,352 @@
+public fun add_partial_key_server(
+    key_server: &mut KeyServer,
+    member: address,
+    partial_pk: vector<u8>,
```
this is called by committee when finalize_committee_for_rotation. i dont think the committee has the url, but we can add it to propose so each member can submit an approval and also its url

joyqvq commented on move/seal_testnet/sources/key_server.move:
```diff
@@ -0,0 +1,352 @@
+    key_server: &mut KeyServer,
+    members: &vector<address>,
+    partial_pks: &vector<vector<u8>>,
+    ctx: &mut TxContext,
```
this is a helper function that the committee.move calls to add all partial key server objects as df to key server's keyserverV2 df. 

do you mean i dont need a helper function? yes but the id field of KeyServerV2 had to be exposed to this pkg, though its easier just to add a helper function do this.

joyqvq commented on move/committee/sources/committee.move:
```diff
@@ -0,0 +1,1034 @@
+    ctx: &mut TxContext,
+) {
+    assert!(committee.members.contains(&ctx.sender()), ENotMember);
+    let mut key_server = transfer::public_receive(&mut committee.id, ks);
```
this call in committee.move calls the the helper function in keyserver.move that only the member that has an entry in the keyserverV2's df. it uses ctx.sender() to look for it. so this only allows you to update your own url if you in the finalized committee. someone not in the committee will throw EPartialKeyServerNotFound
```
/// Update the URL of a partial key server, can only update the caller created server.
public fun update_partial_ks_url(key_server: &mut KeyServer, url: String, ctx: &mut TxContext) {
    assert!(has_v2(key_server), EInvalidVersion);
    let v2: &mut KeyServerV2 = key_server.v2_mut();
    assert!(df::exists_(&v2.id, ctx.sender()), EPartialKeyServerNotFound);
    let partial_key_server: &mut PartialKeyServer = df::borrow_mut(&mut v2.id, ctx.sender());
    partial_key_server.url = url;
}
```

joyqvq commented on move/committee/sources/committee.move:
```diff
@@ -0,0 +1,1109 @@
+/// Add the caller to approvals list. If already in PostDKG state, checks that
+/// submitted partial_pks are consistent with the onchain state, then add the
+/// caller to approvals list.
+public fun propose_for_rotation(
```
currently the propose_for_rotation does not deal with the master PK whereas propose from fresh dkg does.

during finalize_committee_for_rotation, the key server object is specified in function signature, we just move the entire ks obj over without caring at the pk. 

if we want to combine the methods, we would need to save the PK in Finalized state so propose rotation need to read and maintain it, but its one additional thing to store (its already in ks we shouldnt need to care about it for rotation)
